File: javascript/deletion_distance.js
function deletionDistance(str1, str2) {
  var memo = {};

  return dHelper(str1, str2, str1.length, str2.length, memo);
}

function dHelper(str1, str2, i, j, cache) {
  if (i === 0) { return j; }
  if (j === 0) { return i; }

  var cacheKey;

  if (str1[i - 1] === str2[j - 1]) {
    cacheKey = (i - 1) + ',' + (j - 1);
    cache[cacheKey] = cache[cacheKey] ? cache[cacheKey] : dHelper(str1, str2, i - 1, j - 1, cache);

    return cache[(i - 1) + ',' + (j - 1)];
  } else {
    cacheKey = i + ',' + (j - 1);
    cache[cacheKey] = cache[cacheKey] ? cache[cacheKey] : dHelper(str1, str2, i, j - 1, cache);

    cacheKey = (i - 1) + ',' + j;
    cache[cacheKey] = cache[cacheKey] ? cache[cacheKey] : dHelper(str1, str2, i - 1, j, cache);

    return 1 + Math.min(cache[i + ',' + (j - 1)], cache[(i - 1) + ',' + j]);
  }
}

File: ruby/deletion_distance.rb
def deletion_distance(str1, str2)
  helper(str1, str2, str1.size - 1, str2.size - 1, {})
end

def helper(str1, str2, i1, i2, cache)
  return i1 + 1 if i2 < 0
  return i2 + 1 if i1 < 0

  if str1[i1] == str2[i2]
    cache[[i1 - 1, i2 - 1]] ||= helper(str1, str2, i1 - 1, i2 - 1, cache)
    cache[[i1, i2]] = cache[[i1 - 1, i2 - 1]]
  else
    cache[[i1 - 1, i2]] ||= helper(str1, str2, i1 - 1, i2, cache)
    cache[[i1, i2 - 1]] ||= helper(str1, str2, i1, i2 - 1, cache)
    cache[[i1, i2]] = 1 + [cache[[i1 - 1, i2]], cache[[i1, i2 - 1]]].min
  end

  cache[[i1, i2]]
end
